### PR TITLE
[Ready For Review] Virtual machine scale set

### PR DESCRIPTION
**Pull request by anuchandy**
Note : The VMSS part that handle load balancer is not complete as we don't have IBackend and IIboundNatPool available yet.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@anuchandy__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (anuchan). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by hovsepm** [comment]
@azuresdkci add to whitelist

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Comment by anudeepsharma** [comment]
> namespace Microsoft.Azure.Management.V2.Compute

Add License info to all files

---
Refers to: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/ComputeManager.cs:14 in c2f763a. [](commit_id = c2f763a206ea212e6e00e564a8a542e60d7c1ab8, deletion_comment = False)

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/MissingNetworkTypes.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/MissingNetworkTypes.cs]: Try better file names, also remove unused using, put license info

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetExtensionImpl.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            }
+        }
+        public bool? AutoUpgradeMinorVersionEnabled
+        {
+            get
+            {
+                return this.Inner.AutoUpgradeMinorVersion;
+            }
+        }
+        public IDictionary<string,object> PublicSettings
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetExtensionImpl.cs]: space after ,

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetExtensionImpl.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        {
+            get
+            {
+                return this.publicSettings;
+            }
+        }
+        public string PublicSettingsAsJsonString
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetExtensionImpl.cs]: better to either remove it if don't want to implement this, also put blank lines between properties. Install CodeMaid plugin on visualstudio it is awesome.

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    /// <summary>
+    /// The implementation for {@link VirtualMachineScaleSets}.
+    /// </summary>
+    public partial class VirtualMachineScaleSetsImpl  :
+        GroupableResources<IVirtualMachineScaleSet,VirtualMachineScaleSetImpl,VirtualMachineScaleSetInner, IVirtualMachineScaleSetsOperations, ComputeManager>,
+        IVirtualMachineScaleSets
+    {
+        private IStorageManager storageManager;
+        private INetworkManager networkManager;
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs]: blank line before but not after

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+        public PagedList<IVirtualMachineScaleSet> ListByGroup (string groupName)
+        {
+            IPage<VirtualMachineScaleSetInner> firstPage = InnerCollection.List(groupName);
+            var pagedList = new PagedList<VirtualMachineScaleSetInner>(firstPage, (string nextPageLink) =>
+            {
+                return InnerCollection.ListNext(nextPageLink);
+            });
+            return WrapList(pagedList);
+
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs]: remove blank line after return but add before return if you like

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                    },
+                    NetworkProfile = new VirtualMachineScaleSetNetworkProfile
+                    {
+                        NetworkInterfaceConfigurations = new List<VirtualMachineScaleSetNetworkConfigurationInner>()
+                        {
+                            new VirtualMachineScaleSetNetworkConfigurationInner
+                            {
+                                Primary = true,
+                                Name = "primary-nic-cfg",
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs]: declare a constant for primary-nic-cfg in class or function

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                                IpConfigurations = new List<VirtualMachineScaleSetIPConfigurationInner>()
+                                {
+                                    new VirtualMachineScaleSetIPConfigurationInner
+                                    {
+                                        Name = "primary-nic-ip-cfg"
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetsImpl.cs]: declare a constant for primary-nic-ip-cfg

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetSkuImpl.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    /// </summary>
+    public partial class VirtualMachineScaleSetSkuImpl  :
+        Wrapper<VirtualMachineScaleSetSku>,
+        IVirtualMachineScaleSetSku
+    {
+        // TODO: Report bug -> autorest generator is not appendeing 'Inner' to type Microsoft.Azure.Management.Compute.Models.VirtualMachineScaleSetSku
+        private VirtualMachineScaleSetSkuImpl (VirtualMachineScaleSetSku inner) : base(inner)
+        {
+
+            //$ super(inner);
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetSkuImpl.cs]: this class not implemented yet

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetSkuTypes.cs [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+        public override bool Equals (object obj)
+        {
+
+            string value = this.ToString();
+            if (!(obj is VirtualMachineScaleSetSkuTypes)) {
+                return false;
+            }
+
+            if (obj == this) {
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetSkuTypes.cs]: fix indentation

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetSkuTypes.cs [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+            return true;
+            }
+            VirtualMachineScaleSetSkuTypes rhs = (VirtualMachineScaleSetSkuTypes) obj;
+            if (value == null) {
+                return rhs.value == null;
+            }
+            return value.Equals(rhs.value);
```
Comment (anudeepsharma) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetSkuTypes.cs]: to Enhance readability compare either ToString() with ToString or .value with .value, right now comparison is between lhs.ToString() and rhs.value (I got implementation of ToString is to return value, this is to enhance maintainability/readability of code)

**Review by anudeepsharma — APPROVED** [review]
Few comments otherwise approved

**Comment by hovsepm** [comment]
@anuchandy please resolve conflicts with Azure repo

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/Domain/IVirtualMachineExtensionImage.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        string TypeName { get; }
+
+        /// <returns> the name of the virtual machine extension image version this image represents</returns>
+        string VersionName { get; }
+
+        /**
+         * @return the operating system this virtual machine extension image supports
+         */
+        /// <returns>the region in which virtual machine extension image is available</returns>
+        OperatingSystemTypes osType();
```
Comment (anuchandy) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/Domain/IVirtualMachineExtensionImage.cs]: good catch!

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/MissingNetworkTypes.cs (hunk 1 above)
Comment (anuchandy) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/MissingNetworkTypes.cs]: This is a temporary file containing the dummy def for interfaces in the network those are not yet available but required for VMSS. Once those interfaces are available this file will be removed

File: src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetExtensionImpl.cs (hunk 3 above)
Comment (anuchandy) [inline src/ResourceManagement/Compute/Microsoft.Azure.Management.V2.Compute/VirtualMachineScaleSetExtensionImpl.cs]: thanks!

**Comment by anuchandy** [comment]
@anudeepsharma @hovsepm  Resolved conflicts and addressed RCs

**Pull request merged**